arch/cortex-m/src/lib.rs

```diff
@@ -15,6 +15,62 @@ pub mod support;
 pub mod syscall;
 pub mod systick;
 
+/// Trait to encapsulate differences in between Cortex-M variants
+///
+/// This trait contains functions and other associated data (constants) which
+/// differs in between different Cortex-M architecture variants (e.g. Cortex-M0,
+/// Cortex-M4, etc.). It is not designed to be implemented by an instantiable
+/// type and passed around as a runtime-accessible object, but is to be used as
+/// a well-defined collection of functions and data to be exposed to
+/// architecture-dependent code paths. Hence, implementations can use an `enum`
+/// without variants to implement this trait.
+///
+/// The fact that some functions are proper trait functions, while others are
+/// exposed via associated constants is necessitated by the associated const
+/// functions being `#\[naked\]`. To wrap these functions in proper trait
+/// functions would require these trait functions to also be `#\[naked\]` to
+/// avoid generating a function prologue and epilogue, and we'd have to call the
+/// respective backing function from within an asm! block. The associated
+/// constants allow us to simply reference any function in scope and export it
+/// through the provided CortexMSubarch trait infrastructure.
+pub trait CortexMSubarch {
```

Comment:
Subarch is a little confusing, because there is no cortex-m arch. Maybe `CortexMFamily` or `CortexMVariant`.